File: modules/cargo_mission.c
module cargo_mission {
  import ai_stationary;
	object faction;
	object destination;
	object basecontainer;
	object cargoname;
	bool arrived;
	int difficulty;
	float distfrombase;
	int quantity;
	float cred;
	import random;
	import launch;
	import faction_ships;
	bool capship;
	void sys (object currentsystem, int sysaway) {
	  if (sysaway<=0) {
	    destination=currentsystem;
	  } else {
	    int max=_std.getNumAdjacentSystems(currentsystem);
	    if (max>0) {
	      int nextsysnum=random.randomint(0,max-1);
	      object nextsystem=_std.getAdjacentSystem(currentsystem,nextsysnum);
	      sys(nextsystem,sysaway-1);
	    } else {
	      destination="sol_sector/sol";
	    }
	  }
	};
	
	void init (int factionname, int numsystemsaway, int cargoquantity, int missiondifficulty, float distance_from_base, float creds, bool launchoncapship) {
	  capship= launchoncapship;
	  faction_ships.init();
	  faction=faction_ships.intToFaction(factionname);
	  arrived=false;
	  cred=creds;
	  distfrombase=distance_from_base;
	  difficulty=missiondifficulty;
	  object mysys=_std.getSystemFile();
	  quantity=cargoquantity;
	  object sysfile = _std.getSystemFile();
	  sys(sysfile,numsystemsaway);
	  object you=_unit.getPlayer();
	  if (quantity<1){
	    quantity=1;
	  }
	  object list=_unit.getRandCargo(quantity);
	  int tempquantity=quantity;
	  cargoname=_olist.at(list,0);
	  if (!_std.isNull(you)) {
	    quantity = _unit.addCargo(you,cargoname        ,_olist.at(list,1),_olist.at(list,2),_olist.at(list,3),_olist.at(list,4),_olist.at(list,5));  //ADD CARGO HERE
	    if (tempquantity>0) {
	      cred=cred *_std.Float(quantity)/_std.Float(tempquantity);
	    }
	    
	    object str = _string.new();
	    object name = _unit.getName (you);
	    _io.sprintf(str,"Good Day, %s",name);
	    _io.message (0,"game","all",str);
	    _io.sprintf(str,"Go to system %s and give the",destination);
	    _io.message (1,"game","all",str);
	    _io.sprintf(str,"cargo to a unit from our faction: %s",faction);
	    _io.message (2,"game","all",str);
	    _io.sprintf(str,"%d of the %s cargo",quantity,cargoname);
	    _io.message (3,"game","all",str);
	    _string.delete(str);
	    _olist.delete(list);
	  } else {
	    _std.terminateMission (false);
	  }
	};
	void takeCargoAndTerminate (object you) {
	  int removenum=_unit.removeCargo(you,cargoname,quantity,true);
	  if ((removenum==quantity)||(quantity==0)) {
	    _io.message (0,"game","all","Excellent work pilot.");
	    _io.message (0,"game","all","You have been rewarded for your effort as agreed.");
	    _io.message (0,"game","all","Your contribution to the war effort will be remembered.");
	    _unit.addCredits(cred);
	    _std.terminateMission(true);
	  } else {
	    _io.message (0,"game","all","You did not follow through on your end of the deal.");

	    if (difficulty<1) {
	      _io.message (0,"game","all","Your pay will be reduced");
	      _io.message (0,"game","all","And we will consider if we will accept you on future missions.");
	      float addcred=(_std.Float(removenum)/_std.Float((quantity*(1+difficulty))))*cred;
	      _unit.addCredits(addcred);
	    } else {
	      _io.message (0,"game","all","You will not be paid!");
	      if (difficulty>=2) {
		_io.message (0,"game","all","And your idiocy will be punished.");
		_io.message (0,"game","all","You had better run for what little life you have left.");
		int i=0;
		object un;


		while (i<difficulty) {
		  un=faction_ships.getRandomFighter(faction);
		  object newunit=launch.launch_wave_around_unit("shadow", faction, un, "default", 1, 1000.0,you);
		  _unit.setTarget(newunit,you);
		  i=i+1;
		}
	      }
	    }
	    _std.terminateMission(false);

	  }
	};
	void loop () {
	  if (arrived) {
	    object base=_unit.getUnitFromContainer(basecontainer);
	    object you=_unit.getPlayer();
	    if (_std.isNull(base)||_std.isNull(you)) {
	      _std.terminateMission(false);
	      return;
	    }
	    float dist=_unit.getDistance(base,you);
	    if (dist<=distfrombase) {
	      takeCargoAndTerminate(you);		
	      return;
	    }
	  } else {
	    object sysfil = _std.getSystemFile();
	    if (_string.equal (sysfil,destination)) {
	      arrived=true;
	      object newship=faction_ships.getRandomCapitol(faction);
	      int randint=random.randomint(0,50);
	      object significant = unit.getSignificant (randint);
	      if (_std.isNull (significant)) {
		significant =_unit.getPlayer();
	      }else {
		object fgname = _unit.getFgName (significant);
		while (!((_unit.isPlanet(significant))||(_string.equal(fgname,"Base")))&&(randint<1024)) {
		  _string.delete (fgname);
		  randint=randint+1;
		  significant =unit.getSignificant(randint);
		  fgname = _unit.getFgName (significant);
		}
	      }
	      if (_std.isNull(significant)) {
		arrived=false;
	      }else {
		if (capship) {
		  significant=launch.launch_wave_around_unit("Base",faction,newship,"_ai_stationary",1,5000.0,significant);
		}
		object str = _string.new();
		object name = _unit.getName (significant);
		_io.sprintf(str,"You must drop your cargo off with the %s unit",name);
		_io.message (0,"game","all",str);
		_string.delete(str);
		
		basecontainer=_unit.getContainer(significant);
	      }
	    }
	    _string.delete (sysfil);
	  }
	};
}

File: modules/cargo_mission.py
from go_to_adjacent_systems import *
from go_somewhere_significant import *
import vsrandom
import launch
import faction_ships
import VS
import Briefing
import universe
import unit
import Director
import quest
class cargo_mission (Director.Mission):
	def initbriefing(self):
		VS.IOmessage (0,"cargo mission","briefing","Your mission for today will be to run some %s cargo" % self.cargoname)
		self.briefgametime = 0
		self.adjsys.initbriefing()

	
	def loopbriefing(self):
		brief_you=self.adjsys.loopbriefing()
		if (brief_you != -1):
			VS.IOmessage(0,"cargo mission","briefing","Once there, you must drop the cargo off at a specified unit")
			if (self.briefgametime==0):
				self.briefgametime = VS.GetGameTime()
			elif ((VS.GetGameTime()-self.briefgametime)>5):
				Briefing.terminate()
	def endbriefing(self):
		self.adjsys.endbriefing()
		del self.briefgametime
	def SetVar (self,val):
	  if (self.var_to_set!=''):
	    quest.removeQuest (self.you.isPlayerStarship(),self.var_to_set,val)
	def __init__ (self,factionname, numsystemsaway, cargoquantity, missiondifficulty, creds, launchoncapship, time_to_complete, category,jumps=(),var_to_set=''):
	  Director.Mission.__init__(self);
	  self.you=VS.Unit()
	  self.base=VS.Unit()
	  self.arrived=0
	  self.var_to_set=var_to_set
	  self.mplay="all"
#	  self.mission_time=VS.GetGameTime()+time_to_complete*100*float(1+numsystemsaway)
	  self.capship= launchoncapship
	  self.faction=factionname
	  self.cred=creds
	  self.difficulty=missiondifficulty
	  self.you=VS.getPlayer()
	  self.adjsys=go_to_adjacent_systems(self.you,numsystemsaway,jumps)
	  self.quantity=cargoquantity
	  self.mplay=universe.getMessagePlayer(self.you)
	  if (self.quantity<1):
	    self.quantity=1
	  carg=VS.getRandCargo(self.quantity,category)
	  if (carg.GetQuantity()==0):
	    carg = VS.getRandCargo(self.quantity,"") #oh no... could be starships...
	  tempquantity=self.quantity
	  self.cargoname=carg.GetContent()
	  name = self.you.getName ()
	  carg.SetMissionFlag(1)
	  if (not self.you.isNull()):
	    self.quantity = self.you.addCargo(carg)  #I add some cargo
	  else:
	    VS.IOmessage (2,"cargo mission",self.mplay,"#ff0000Unable to establish communications. Mission failed.")
	    VS.terminateMission (0)
	    return
#	  creds_deducted = (carg.GetPrice()*float(self.quantity)*vsrandom.random()+1)
#	  self.cred += creds_deducted
	  if (tempquantity>0):
	    self.cred*=float(self.quantity)/float(tempquantity)
	  else:
	    VS.IOmessage (2,"cargo mission",self.mplay,"#ff0000You do not have space to add our cargo to the mission. Mission failed.")
	    VS.terminateMission(0)
	    return
	  
	  if (self.quantity==0):
	    VS.IOmessage (2,"cargo mission",self.mplay,"#ff0000You do not have space to add our cargo to the mission. Mission failed.")
	    VS.terminateMission(0)
	    return
	  
	  VS.IOmessage (0,"cargo mission",self.mplay,"Good Day, %s. Your mission is as follows:" % (name))
	  self.adjsys.Print("You should start in the system named %s","Then jump to %s","Finally, jump to %s, your final destination","cargo mission",1)
	  VS.IOmessage (2,"cargo mission",self.mplay,"Give the cargo to a %s unit." % (self.faction))
	  VS.IOmessage (3,"cargo mission",self.mplay,"You will receive %d of the %s cargo" % (self.quantity,self.cargoname))
#	  VS.IOmessage (4,"cargo mission",self.mplay,"We will deduct %.2f credits from your account for the cargo needed." % (creds_deducted))
	  VS.IOmessage (4,"cargo mission",self.mplay,"You will earn %.2f credits when you deliver our cargo." % (creds))
	  VS.IOmessage (4,"cargo mission",self.mplay,"#00ff00Good luck!")
#	  self.you.addCredits (-creds_deducted)
	
	def takeCargoAndTerminate (self,you, remove):
	  removenum=0 #if you terminate without remove, you are SKREWED
	  if (remove):
	    removenum=you.removeCargo(self.cargoname,self.quantity,1)
	    print "removed %d" % removenum
	    mpart=VS.GetMasterPartList()
	    newcarg=mpart.GetCargo(self.cargoname)
	    newcarg.SetQuantity(removenum)
	    self.base.addCargo(newcarg)
	    has=self.you.hasCargo(self.cargoname)
	    if (has):
	      has=self.you.removeCargo(self.cargoname,has,0)
	      newcarg.SetMissionFlag(0)
	      newcarg.SetQuantity(has)
	      self.you.addCargo(newcarg) #It seems that removing and then adding it again is the only way...
	  if ((removenum>=self.quantity) or (self.quantity==0)):
	    VS.IOmessage (0,"cargo mission",self.mplay,"#00ff00Excellent work pilot.")
	    VS.IOmessage (0,"cargo mission",self.mplay,"#00ff00You have been rewarded for your effort as agreed.")
	    VS.IOmessage (0,"cargo mission",self.mplay,"#00ff00Your excellent work will be remembered.")
	    you.addCredits(self.cred)
	    self.SetVar(1)
	    VS.terminateMission(1)
	    return
	  else:
	    VS.IOmessage (0,"cargo mission",self.mplay,"#ff0000You did not follow through on your end of the deal.")
	    if (self.difficulty<1):
	      VS.IOmessage (0,"cargo mission",self.mplay,"#ff0000Your pay will be reduced")
	      VS.IOmessage (0,"cargo mission",self.mplay,"#ff0000And we will consider if we will accept you on future missions.")
	      addcred=(float(removenum)/(float(self.quantity*(1+self.difficulty))))*self.cred
	      you.addCredits(addcred)
	    else:
	      VS.IOmessage (0,"cargo mission",self.mplay,"#ff0000You will not be paid!")
	      universe.punish(self.you,self.faction,self.difficulty)
	    self.SetVar(-1)
	    VS.terminateMission(0)
	    return
	  
	
	def Execute (self):
##	  if (VS.getGameTime()>mission_time):
##	    VS.IOmessage (0,"cargo mission",self.mplay,"You Have failed to deliver your cargo in a timely manner.")
##	    VS.IOmessage (0,"cargo mission",self.mplay,"The cargo is no longer of need to us.")
##	    if (you):
##	      takeCargoAndTerminate(you,0)
##	    return
	  if (self.you.isNull() or (self.arrived and self.base.isNull())):
	    VS.IOmessage (0,"cargo mission",self.mplay,"#ff0000You were unable to deliver cargo. Mission failed.")
	    VS.terminateMission(0)
	    return
	  if (not self.adjsys.Execute()):
	    return
	  if (self.arrived):
	    self.adjsys.Execute=self.adjsys.HaveArrived
	    if (self.base.isDocked(self.you)):
	      self.takeCargoAndTerminate(self.you,1)		
	      return
	  else:
	    self.arrived=1
	    self.adjsys=go_somewhere_significant(self.you,1,100,self.capship,self.faction)
	    capstr="planet"
	    dockstr="land"
	    if (self.capship):
	      dockstr="dock"
	      capstr="ship"
	    self.adjsys.Print("You must visit the %%s %s" % (capstr),"cargo mission",", docked around the %s",0)
	    VS.IOmessage(0,"cargo mission",self.mplay,"Once there, %s and we will transport the cargo off of your ship." % (dockstr))
	    self.base=self.adjsys.SignificantUnit()
	    self.base.SetHull(10000000000.00)
def initrandom (factionname, missiondifficulty,creds_per_jump, launchoncapship, sysmin, sysmax, time_to_complete, category,jumps=(),var_to_set=''):
	numsys=vsrandom.randrange(sysmin,sysmax)
	return cargo_mission(factionname,numsys, vsrandom.randrange(4,15), missiondifficulty,creds_per_jump*float(1+numsys),launchoncapship, 10.0, category,jumps,var_to_set)

